# model_evaluation exercises

## Given the following confusion matrix, evaluate (by hand) the model's performance.
|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

### 1. In the context of this problem, what is a false positive?
    - FP = actual dog and predicted cat

### 2. In the context of this problem, what is a false negative?
    - FN = actual cat and predicted cat

### How would you describe this model?
    - This chart looks slightly different in that the predictions are the columns and the actual are the rows.

In [1]:
import pandas as pd

In [2]:
data = {'actual_dog':['46','7'],'actual_cat':['13','34']}
df = pd.DataFrame(data, columns = ['actual_dog','actual_cat'], index=['pred_dog','pred_cat'])
df # a better way to read the data

actual_dog actual_cat
pred_dog         46         13
pred_cat          7         34

## You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

### Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

### Use the predictions dataset and pandas to help answer the following questions:

    - An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
    
    - Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [8]:
import csv

In [10]:
df = pd.read_csv('c3.csv')

In [11]:
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [13]:
# They tell you that they want to identify as many of the ducks that have a defect as possible. 
# Which evaluation metric would be appropriate here? 
# Which model would be the best fit for this use case?

pd.crosstab(df.actual, df.model1) # looking at model 1

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

In [14]:
pd.crosstab(df.actual, df.model2) # looking at model 2

model2     Defect  No Defect
actual                      
Defect          9          7
No Defect      81        103

In [15]:
pd.crosstab(df.actual, df.model3) # looking at model 3

model3     Defect  No Defect
actual                      
Defect         13          3
No Defect      86         98

In [16]:
df.actual.value_counts() # looking at counts of Defects and No Defects

No Defect    184
Defect        16
Name: actual, dtype: int64

- TP = Actual "Defect" and Predicted "Defect"
- TN = Actual "No Defect" and Predicted "Defect"
- FP = Actual "Defect" and Predicted "No Defect"
- FN = Actual "No Defect" and Predicted "No Defect"

In [17]:
df['baseline_prediction'] = 'No Defect' # No defect is most frequent, adding a baseline of no defect

In [19]:
df.head() # check_yo_head

actual     model1  model2     model3 baseline_prediction
0  No Defect  No Defect  Defect  No Defect           No Defect
1  No Defect  No Defect  Defect     Defect           No Defect
2  No Defect  No Defect  Defect  No Defect           No Defect
3  No Defect     Defect  Defect     Defect           No Defect
4  No Defect  No Defect  Defect  No Defect           No Defect

In [33]:
# we want to identify the defect ducks
# use recall for model1

subset = df[df.actual == 'Defect']

model_recall = (subset.model1 == subset.actual).mean()
baseline_recall = (subset.baseline_prediction == subset.actual).mean()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model recall: 50.00%
baseline recall: 0.00%


In [34]:
# we want to identify the defect ducks
# use recall for model2

subset = df[df.actual == 'Defect']

model_recall = (subset.model2 == subset.actual).mean()
baseline_recall = (subset.baseline_prediction == subset.actual).mean()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model recall: 56.25%
baseline recall: 0.00%


In [35]:
# we want to identify the defect ducks
# use recall for model3

subset = df[df.actual == 'Defect']

model_recall = (subset.model3 == subset.actual).mean()
baseline_recall = (subset.baseline_prediction == subset.actual).mean()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model recall: 81.25%
baseline recall: 0.00%


## Which model would be the best fit for this use case?

- Model3 
- model recall: 81.25%

In [36]:
# df.replace(['No Defect', 'Defect'],  [0, 1])
# my code above was not running and I thought it was because it was not 0,1 but not needed
# make sure to check capitalization 'defect' vs 'Defect'


## They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. 

- Which evaluation metric would be appropriate here? 
- Which model would be the best fit for this use case?


In [38]:
df.head() # check_yo_head

actual     model1  model2     model3 baseline_prediction
0  No Defect  No Defect  Defect  No Defect           No Defect
1  No Defect  No Defect  Defect     Defect           No Defect
2  No Defect  No Defect  Defect  No Defect           No Defect
3  No Defect     Defect  Defect     Defect           No Defect
4  No Defect  No Defect  Defect  No Defect           No Defect

In [42]:
subset = df[df.actual == 'Defect']

model_precision = (subset.actual == subset.model3).mean()
print(f'model precision: {model_precision:.2%}')

model precision: 81.25%


In [43]:
model_precision

0.8125

In [48]:
# Baseline Precision: # only look at the subset where we made positive prediction (i.e. prediction == 'coffee')
# there is no 'Defect' in the baseline_prediction column and therefore unble to do the calculatio
# returns not a number 'nan'
subset = df[df.baseline_prediction == 'Defect']

baseline_precision = (subset.actual == subset.baseline_prediction).mean()
print(f'baseline precision: {baseline_precision:.2%}')

baseline precision: nan%


# 4 You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

## At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

### Several models have already been developed with the data, and you can find their results here.

## Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions

- a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

- b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

- c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?